debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerStreamingChangeEventSource.java

```diff
@@ -447,4 +458,34 @@ private Lsn getToLsn(SqlServerConnection connection, String databaseName, TxLogP
 
         return connection.getNthTransactionLsnFromLast(databaseName, fromLsn, maxTransactionsPerIteration);
     }
+
+    @Override
+    public void commitOffset(Map<String, ?> offset) {
+        Lsn commitLsn = Lsn.valueOf((String) offset.get("commit_lsn"));
+        synchronized (streamingExecutionContexts) {
+            for (Map.Entry<SqlServerPartition, SqlServerStreamingExecutionContext> entry : streamingExecutionContexts.entrySet()) {
+                SqlServerPartition partition = entry.getKey();
+                List<SqlServerChangeTable> changeTablesWithKnownStopLsn = entry.getValue().getChangeTablesWithKnownStopLsn();
+
+                synchronized (changeTablesWithKnownStopLsn) {
+                    List<SqlServerChangeTable> changeTablesToBeDeleted = changeTablesWithKnownStopLsn.stream()
+                            .filter(t -> t.getStopLsn().compareTo(commitLsn) < 0)
+                            .collect(Collectors.toList());
+
+                    for (SqlServerChangeTable table : changeTablesToBeDeleted) {
+                        try {
+                            dataConnection.deleteChangeTable(partition, table);
+                        }
+                        catch (SQLException e) {
+                            throw new RuntimeException(e);
+                        }
+                        LOGGER.info("Deleted change table {} as the committed change lsn ({}) is greater than the table's stop lsn", table, offset);
+
+                    }
+
+                    changeTablesWithKnownStopLsn.removeAll(changeTablesToBeDeleted);
```

Comment:
What will happen if one of the above "delete" operations succeeds but another fails? It looks like we will have the first capture instance deleted but the offset won't be committed, so the connector will try using the deleted capture instance next time.

Either way, let's think about what can go wrong here and how to make this process idempotent.

Reply:
I'll test what happens when `commitOffset` throws an exception. We need to attempt to delete capture instances only after the commit has happened.

Also we could just log an error and let `commitOffset` finish without exceptions. It'll retry to delete capture instances on the next `commitOffset` call.

Reply:
> We need to attempt to delete capture instances only after the commit has happened.

This sounds right. Additionally, do we even need to fail the task if an attempt to delete a capture instance failed? Maybe it could be a retriable exception? This way, the task will internally restart, get the list of capture instances and attempt to delete the outdated one again.

Reply:
Jira ticket: rework the removal of capture instances from the map by removing them one by one upon successful deletion from the database.